### PR TITLE
Backport fixes from the master branch to the v7 branch

### LINKED ISSUE #92: Range miscalculation when using If expressions
```elm
module TestModule exposing (..)

a =
  if cond then
    1
  else
    2
  
-- hello

{-| doc
-}
b = 3
```

In the code above, the range for `a` ends where `b` starts, meaning it will include the comment in between and the documentation for `b`.

If the if expression is removed, like here:
```elm
a =
  2
  
-- hello

{-| doc
-}
b = 3
```

then `a`'s range seems to be correct.

This looks very similar to https://github.com/stil4m/elm-syntax/issues/63 which had the same problem but for let expressions.

### PR DESCRIPTION
**Pull request by jfmengels**
Before releasing a new major version, I would like to publish a new v7 version with all the fixes currently present on `master`, and all the issues that have currently been reported.

This PR contains cheer-picks of everything that didn't seem to be meant to change the API of the package. The only remaining thing is to remove the breaking change introduced by fce06d96583906f84c863ce8b48bd8e6adc2e921 (EDIT: It's done), although it might be interesting to fix #92, because that sounds very related, before undoing the breaking change.

### CODE REVIEW

File: src/Elm/Syntax/Exposing.elm [hunk 1]
```diff
@@ -58,7 +58,7 @@ type TopLevelExpose
 {-| Exposed Type
 -}
 type alias ExposedType =
-    { name : String
+    { name : Node String
```
Comment (MartinSStewart) [inline src/Elm/Syntax/Exposing.elm:61]: Isn't this a breaking change?

Comment (jfmengels) [reply src/Elm/Syntax/Exposing.elm:61]: Yes, that is the breaking change I mentioned in the description. I had a quick look at removing it, but I'm not yet familiar enough with the code and didn't succeed. I'll try again later.

**Comment by jfmengels** [comment]
Alright, I was able to remove the breaking change (much easier after a good night's sleep :smile: ) :+1:

File: src/Elm/Syntax/Exposing.elm (hunk 1 above)
Comment (MartinSStewart) [reply src/Elm/Syntax/Exposing.elm]: Oops, my bad. Easier to read PRs after a good nights sleep as well!

**Pull request merged**